yarn-project/noir-private-kernel/src/crates/private-kernel-lib/src/hash.nr

```diff
@@ -0,0 +1,259 @@
+use crate::address::{Address, EthAddress};
+use crate::mocked::VerificationKey;
+use crate::point::Point;
+use crate::abis::function_selector::FunctionSelector;
+use crate::abis::function_leaf_preimage::FunctionLeafPreimage;
+use crate::abis::new_contract_data::NewContractData as ContractLeafPreimage;
+use crate::abis::function_data::FunctionData;
+use crate::utils::uint128::U128;
+use crate::utils::uint256::U256;
+use crate::utils::bounded_vec::BoundedVec;
+
+use dep::aztec::{
+    constants_gen,
+    constants_gen::{CONTRACT_TREE_HEIGHT, FUNCTION_TREE_HEIGHT, NOTE_HASH_TREE_HEIGHT},
+    hash::sha256_to_field,
+};
+
+// Calculate the Merkle tree root from the sibling path and leaf.
+//
+// The leaf is hashed with its sibling, and then the result is hashed
+// with the next sibling etc in the path. The last hash is the root.
+//
+// TODO(David/Someone): The cpp code is using a uint256, whereas its
+// TODO a bit simpler in Noir to just have a bit array.
+// TODO: I'd generally like to avoid u256 for algorithms like 
+// this because it means we never even need to consider cases where 
+// the index is greater than p.
+fn root_from_sibling_path<N, T>(leaf : Field, leaf_index : T, to_bits : fn(T) -> [u1; N], sibling_path : [Field; N]) -> Field {
+    let mut node = leaf;
+    let indices = to_bits(leaf_index);
+
+    for i in 0..N {
+        // indices[i]; // This line will fail.
+        if indices[i] == 1 {
+            node = merkle_hash(sibling_path[i], node);
+        } else {
+            node = merkle_hash(node, sibling_path[i]);
+        }
+    }
+    node
+}
+
+fn u4_to_le_bit_array(value : u4) -> [u1;4] {
+    let mut le_bit_array : [u1; 4] = [0; 4];
+    for i in 0..4 {
+        le_bit_array[i] = ((value >> i) & 1) as u1;
+    }
+    le_bit_array
+}
+fn u16_to_le_bit_array(value : u16) -> [u1;16] {
+    let mut le_bit_array : [u1; 16] = [0; 16];
+    for i in 0..16 {
+        le_bit_array[i] = ((value >> i) & 1) as u1;
+    }
+    le_bit_array
+}
+fn u32_to_le_bit_array(value : u32) -> [u1;32] {
+    let mut le_bit_array : [u1; 32] = [0; 32];
+    for i in 0..32 {
+        le_bit_array[i] = ((value >> i) & 1) as u1;
+    }
+    le_bit_array
+}
```

Comment:
```suggestion
fn u4_to_le_bit_array(value : u4) -> [u1;4] {
    let mut le_bit_array : [u1; 4] = [0; 4];
    for i in 0..4 {
        le_bit_array[i] = (value >> i) as u1;
    }
    le_bit_array
}
fn u16_to_le_bit_array(value : u16) -> [u1;16] {
    let mut le_bit_array : [u1; 16] = [0; 16];
    for i in 0..16 {
        le_bit_array[i] = (value >> i) as u1;
    }
    le_bit_array
}
fn u32_to_le_bit_array(value : u32) -> [u1;32] {
    let mut le_bit_array : [u1; 32] = [0; 32];
    for i in 0..32 {
        le_bit_array[i] = (value >> i) as u1;
    }
    le_bit_array
}
```

Casting to a `u1` performs the truncation we're doing with the `& 1` implicitly so we can remove it. This takes us from 272306 to 235186 constraints.